---
title: "Assign taxonomy, customize alignment database, align sequences, and quality control"
output: html_document
---

##Assign taxonomy to OTUs

```{r, engine='bash', results='hide'}
#Decompress Greengenes database: gg_13_8_otus
tar -xzvf gg_13_8.gz
```

```{r, engine='bash'}
#Assing taxonomy using Greengenes database (gg_13_8_otus) as reference sequences
assign_taxonomy.py -i usearch_outputs/otus2.fa -t gg_13_8/97_otu_taxonomy.txt -r gg_13_8/97_otus.fasta -m mothur -o assigned_gg_taxa
```

```{r, engine='bash'}
#Add taxonomy data to OTU table with column header "taxonomy" and convert OTU table to biom format
mkdir biom_files

awk 'NR==1; NR > 1 {print $0 | "sort"}' otu_tables/otu_table.txt > assigned_gg_taxa/otu_table.sort.txt 

sort assigned_gg_taxa/otus2_tax_assignments.txt > assigned_gg_taxa/otus2_tax_assignments.sort.txt

{ printf '\ttaxonomy\t\n'; cat assigned_gg_taxa/otus2_tax_assignments.sort.txt ; }  > assigned_gg_taxa/otus2_tax_assignments_sort_label.txt

paste assigned_gg_taxa/otu_table.sort.txt <(cut -f 2 assigned_gg_taxa/otus2_tax_assignments_sort_label.txt) > biom_files/otu_table_tax.txt

biom convert -i biom_files/otu_table_tax.txt -o biom_files/otu_table_tax.biom --to-json --table-type "OTU table" --process-obs-metadata taxonomy 
```

##Customize Silva database to V4 region for sequences alignment

```{r, engine='bash'}
#Create a folder to store Silva database
mkdir silva

#Download and decompress Silva database
wget http://www.mothur.org/w/images/b/be/Silva.nr_v123.tgz -P silva

tar -zxvf silva/Silva.nr_v123.tgz -C silva

#Customize reference database to V4 region
mothur "#pcr.seqs(fasta=silva/silva.nr_v123.align, start=11894, end=25319, keepdots=F, processors=8)"

#Check for outliers
mothur "#summary.seqs(fasta=silva/silva.nr_v123.pcr.align, processors=8)"

#Remove outliers
mothur "#screen.seqs(fasta=silva/silva.nr_v123.pcr.align, summary=silva/silva.nr_v123.pcr.summary, start=1, end=13425, processors=8)"

#Rename customized reference alignment 
mothur "#system(mv silva/silva.nr_v123.pcr.good.align silva/silva.v4.fasta)"
```

##Align sequences

```{r, engine='bash'}
#Align sequences
mothur "#align.seqs(fasta=usearch_outputs/otus2.fa, reference=silva/silva.v4.fasta, processors=8)"

#Summarize quality of alignment
mothur "#summary.seqs(fasta=usearch_outputs/otus2.align, processors=8)"
```

##Quality control of OTUs

Make a file containing all OTUs that aligned properly

```{r}
dir.create("filter_files")

otus_aligned <- read.table("usearch_outputs/otus2.summary", header = T, sep = "\t", stringsAsFactors = F)

otus_correct_aligned <- subset(otus_aligned, start >= 1960 & end == 11550)

write.table(otus_correct_aligned$seqname, file = "filter_files/otus_correct_aligned.txt", col.names = F, row.names = F, quote = F)
```

Remove OTUs with poor alignment, sinlgetons, and classified within the Archaea kingdom or Cyanobacteria phylum.

```{r, engine='bash'}
#Filter out OTUs with poor alignment and singletons
filter_otus_from_otu_table.py -i biom_files/otu_table_tax.biom -o biom_files/otu_table_tax_aligned.biom -e filter_files/otus_correct_aligned.txt --negate_ids_to_exclude -n 2

##Filter out OTUs classified within the Archaea kingdom or Cyanobacteria phylum
filter_taxa_from_otu_table.py -i biom_files/otu_table_tax_aligned.biom -o biom_files/otu_table_final.biom -n k__Archaea,p__Cyanobacteria
```